### PR TITLE
fix: ROX-21573, wait for gathering goroutines to terminate in tests

### PR DESCRIPTION
**Pull request by 0x656b694d**
## Description

As the test gathering procedure calls `Stop()`, there was a race condition between cancelling the context and calling `Track()`.
The first is read by the test as a way to wait for the termination, before checking the result, while the `Track()` call could happen after the check.

The PR adds a wait group, which can be used in tests to ensure all gathering goroutines are terminated before checking the outcome.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~Evaluated and added CHANGELOG entry if required~
- [ ] ~Determined and documented upgrade steps~
- [ ] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

### Here I tell how I validated my change

I added a `time.Sleep(time.Second)` call to the `identify()` function, and so reproduced the race condition problem.
I added `Wait()` calls for the `procs` WaitGroup in the test to ensure the `loop` and `identify` goroutines are terminated before checking the `Track()` calls.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: pkg/telemetry/phonehome/gatherer.go [hunk 1]
```diff
@@ -83,21 +87,26 @@ func (g *gatherer) identify() {
...
+			g.procs.Add(1)
 			go g.identify()
 		case <-g.stopSig.Done():
-			ticker.Stop()
-			return
+			break for_loop
 		}
 	}
+	ticker.Stop()
+	g.procs.Done()
```
Comment (vikin91) [inline pkg/telemetry/phonehome/gatherer.go:109]: This could be in a deferred statement

File: pkg/telemetry/phonehome/gatherer.go [hunk 2]
```diff
@@ -83,21 +87,26 @@ func (g *gatherer) identify() {
...
+for_loop:
 	for !g.stopSig.IsDone() {
 		select {
 		case <-ticker.C:
+			g.procs.Add(1)
 			go g.identify()
 		case <-g.stopSig.Done():
-			ticker.Stop()
-			return
+			break for_loop
```
Comment (vikin91) [inline pkg/telemetry/phonehome/gatherer.go:105]: This could be a simple `return` if you'd use the deferred statement for `ticker.Stop(); g.procs.Done()`

**Review by vikin91 — APPROVED** [review]
Thanks! Let's see what those unit tests do in CI :)

**Comment by 0x656b694d** [comment]
Supersedes #9271.

**Review by dhaus67 — COMMENTED** [review]
A bit late to the party, but the changes done now overcomplicate things when the test code can be adjusted to ensure this doesn't happen:
The track calls will be missed sometimes due to the gathering function calling `g.Stop()`, which then leads to the context being done.

Instead, we could simply call `g.Stop` after a pre-defined time with `time.AfterFunc` or our own `concurrency.AfterFunc`; this way we don't have the current wait group changes in the code which seem unnecessary (and also sort-of misuses the initial concept of wait groups, since more than only the main go routine is adding to the waitgroup, including synchronous calls).

**Comment by 0x656b694d** [comment]
While I'm trying to avoid by all means any sleeps and timers in unit tests, the idea about `AfterFunc` made me think of the `Do()` method of the test framework.
Please see #9286 with the refactoring.

**Pull request merged**